From 758cfc42650593b8a1336a9c2fabe08e7bcfe87d Mon Sep 17 00:00:00 2001
From: "Yue, VincentX" <vincentx.yue@intel.com>
Date: Mon, 16 Oct 2017 18:35:21 +0800
Subject: [PATCH] [WA]Fix crash when click open image/audio/video files in
 DocumentUI.

Due to google don't consider use right mouse button to
open files,lead to can't be open.
In the AbstractActionHandler,it have a interface to open file,
but it need a parameter with object DocumentDetails.
And the AbstractActionHandler achieve only need the DocumentDetails's
mode id. So, making a object Details to transmit the mode id.

Change-Id: I7bf3924d7f1ccd929a9fa9250eb2e68fa32ce528
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-50633
Signed-off-by: Yue, VincentX <vincentx.yue@intel.com>
---
 src/com/android/documentsui/base/DocumentInfo.java |  2 +
 .../documentsui/dirlist/DirectoryFragment.java     | 52 ++++++++++++++++++++++
 2 files changed, 54 insertions(+)

diff --git a/src/com/android/documentsui/base/DocumentInfo.java b/src/com/android/documentsui/base/DocumentInfo.java
index 5d9a93c..810fc90 100644
--- a/src/com/android/documentsui/base/DocumentInfo.java
+++ b/src/com/android/documentsui/base/DocumentInfo.java
@@ -292,6 +292,8 @@ public class DocumentInfo implements Durable, Parcelable {
         return (flags & Document.FLAG_DIR_PREFERS_LAST_MODIFIED) != 0;
     }
 
+    public String getDocumentId() {return documentId;}
+
     @Override
     public int hashCode() {
         return derivedUri.hashCode() + mimeType.hashCode();
diff --git a/src/com/android/documentsui/dirlist/DirectoryFragment.java b/src/com/android/documentsui/dirlist/DirectoryFragment.java
index 9096f63..0fcb814 100644
--- a/src/com/android/documentsui/dirlist/DirectoryFragment.java
+++ b/src/com/android/documentsui/dirlist/DirectoryFragment.java
@@ -47,6 +47,7 @@ import android.support.v7.widget.GridLayoutManager.SpanSizeLookup;
 import android.support.v7.widget.RecyclerView;
 import android.support.v7.widget.RecyclerView.RecyclerListener;
 import android.support.v7.widget.RecyclerView.ViewHolder;
+import android.text.TextUtils;
 import android.util.Log;
 import android.util.SparseArray;
 import android.view.ContextMenu;
@@ -739,6 +740,14 @@ public class DirectoryFragment extends Fragment implements SwipeRefreshLayout.On
 
         // Model must be accessed in UI thread, since underlying cursor is not threadsafe.
         List<DocumentInfo> docs = mModel.getDocuments(selected);
+
+        if (docs.size() == 1) {
+            Details detail = new Details(docs.get(0).getDocumentId());
+            mActions.openDocument(detail, ActionHandler.VIEW_TYPE_PREVIEW,
+                    ActionHandler.VIEW_TYPE_REGULAR);
+            return;
+        }
+
         if (docs.size() > 1) {
             mActivity.onDocumentsPicked(docs);
         } else {
@@ -1067,6 +1076,49 @@ public class DirectoryFragment extends Fragment implements SwipeRefreshLayout.On
         });
     }
 
+    private class Details implements DocumentDetails {
+
+        private int mPosition;
+        private String mModelId;
+        private boolean mInSelectionHotspot;
+        private boolean mInDragHotspot;
+
+        public Details(String modeID) {
+            mModelId = modeID;
+        }
+
+        @Override
+        public boolean hasModelId() {
+            return !TextUtils.isEmpty(mModelId);
+        }
+
+        @Override
+        public String getModelId() {
+            return mModelId;
+        }
+
+        @Override
+        public int getAdapterPosition() {
+            return mPosition;
+        }
+
+        @Override
+        public boolean isInSelectionHotspot(InputEvent event) {
+            return mInSelectionHotspot;
+        }
+
+        @Override
+        public boolean isInDragHotspot(InputEvent event) {
+            return mInDragHotspot;
+        }
+
+        @Override
+        public boolean isOverDocIcon(InputEvent event) {
+            return false;
+        }
+
+    }
+
     private final class ModelUpdateListener implements EventListener<Model.Update> {
 
         @Override
-- 
1.9.1

